docs/design/design-discovery.md

```diff
@@ -0,0 +1,100 @@
+
+                          |
+                          |
+                          |  /------------  Relayed
+                      Y:y | /               Address
+                      +--------+
+                      |        |
+                      |  TURN  |
+                      | Server |
+                      |        |
+                      +--------+
+                          |
+                          |
+                          | /------------  Server
+                   X1':x1'|/               Reflexive
+                    +------------+         Address
+                    |    NAT     |
+                    +------------+
+                          |
+                          | /------------  Local
+                      X:x |/               Address
+                      +--------+
+                      |        |
+                      | Agent  |
+                      |        |
+                      +--------+
+
+```
+
+Figure 1. *Candidate Types*
+
+### Current and Future Discovery Plans
+
+Disclaimer: current peer discovery is still in an early POC state. 
+
+**Nodes without a firewall and/or NAT device between them (currently supported):**
+
+- Apex will peer end nodes directly that share the same STUN IP address. 
+- This means when both nodes share the same public or "reflexive address" (as defined in the above figure) then Apex makes an assumption that other peers are likely to have direct access to one another. Each peer has their stun/reflexive address in the peer listing received from the controller.
+- Next the Apex agent will look up the "Local Address" (see Figure 1) of a peer candidate in the peer listing and attempt to probe for connectivity. If this probing succeeds, we consider it a likely candidate match and both peers setup the connection to one another with a /32 host route and wireguard tunnel.
+
+**Nodes with a firewall and/or NAT device between them (currently in progress):**
+
+```
+                               +---------+
+             +--------+        |Signaling|         +--------+
+             | STUN   |        |Server   |         | STUN   |
+             | Server |        +---------+         | Server |
+             +--------+       /           \        +--------+
+                             /             \
+                            /               \
+                           / <- Signaling -> \
+                          /                   \
+                   +--------+               +--------+
+                   |  NAT   |               |  NAT   |
+                   +--------+               +--------+
+                      /                             \
+                     /                               \
+                 +-------+                       +-------+
+                 | Agent |                       | Agent |
+                 |   L   |                       |   R   |
+                 +-------+                       +-------+
+```
```

Comment:
Do you want to show something about how the wireguard traffic flows in this case? I see the `Signaling` label, but it's not clear that the data traffic flows any differently. That seems like an important distinction.

Reply:
Renamed to relay and changed the ASCII art.